src/Utils.php

```diff
@@ -1021,4 +1027,81 @@ public function hasMultipleValues($element) {
     return FALSE;
   }
 
+
+  /**
+   * @inheritDoc
+   */
+  public function checksumUserAccess($c, $cid) {
+    $request = $this->requestStack->getCurrentRequest();
+    $urlCidN = $urlChecksumN = NULL;
+    $session = \CRM_Core_Session::singleton();
+    $urlCid1 = $request->query->get('cid');
+    $urlChecksum1 = $request->query->get('cs');
+
+    $urlCidN = $request->query->get("cid$c");
+    $urlChecksumN = $request->query->get("cs$c");
+
+    $cs = NULL;
+    if ($c == 1 && !empty($urlChecksum1)) {
+      $cs = $urlChecksum1;
+    }
+    elseif (!empty($urlChecksumN)) {
+      $cs = $urlChecksumN;
+    }
+    if ($cs && (($c == 1 && $urlCid1 == $cid) || $urlCidN == $cid)) {
+      $check_access = $this->wf_civicrm_api4('Contact', 'validateChecksum', [
+        'contactId' => $cid,
+        'checksum' => $cs,
+      ])[0] ?? [];
+      if ($check_access['valid']) {
+        if ($c == 1) {
+          $session->set('userID', $cid);
+        }
+        return TRUE;
+      }
+    }
+    // If access is checked for non primary contact, check if c1 has access to view it.
+    elseif ($c != 1 && $this->isContactAccessible($cid)) {
+      return TRUE;
+    }
+    // If no checksum is passed and user is anonymous, reset prev checksum session values if any.
+    if (\Drupal::currentUser()->isAnonymous() && $session->get('userID') && $c == 1 && empty($urlChecksum1)) {
+      $session->reset();
+    }
+    return FALSE;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function isContactAccessible($cid) {
+    $access = $this->wf_civicrm_api4('Contact', 'checkAccess', [
+      'action' => 'get',
+      'values' => [
+        'id' => $cid,
+      ],
+    ], 0);
+    if (!empty($access['access'])) {
+      return TRUE;
+    }
+
+    $request = $this->requestStack->getCurrentRequest();
+    $urlCid1 = $request->query->get('cid') ?? $request->query->get('cid1') ?? NULL;
+    $urlChecksum1 = $request->query->get('cs') ?? $request->query->get('cs1') ?? NULL;
+
+    if (!empty($urlChecksum1) && !empty($urlCid1)) {
+      $valid = $this->wf_civicrm_api4('Contact', 'validateChecksum', [
+        'contactId' => $urlCid1,
+        'checksum' => $urlChecksum1,
+      ])[0] ?? [];
+      if ($valid['valid']) {
+        // checkAccess v4 api does not check for access via relationship.
+        if (\CRM_Contact_BAO_Contact_Permission::allow($cid)) {
```

Comment:
Currently, the checkAccess API or the contact get api in core does not allow to view the contact if its permitted only by relationship. It only checks for contact allowed via ACL - https://github.com/civicrm/civicrm-core/blob/master/CRM/Contact/BAO/Contact/Permission.php#L361. Raised https://lab.civicrm.org/dev/core/-/issues/4610 for additional thoughts.

Reply:
Thank you @jitendrapurohit